keras/layers/crf.py

```diff
@@ -0,0 +1,481 @@
+import keras
+from keras.layers import Layer, InputSpec
+from keras import initializations, activations, regularizers
+import keras.backend as K
+from keras.objectives import categorical_crossentropy, sparse_categorical_crossentropy
+
+if K._BACKEND == 'tensorflow':
+    import tensorflow as tf
+else:
+    import theano
+    import theano.tensor as T
+
+
+class CRF(Layer):
+    '''An implementation of linear chain conditional random field (CRF).
+
+    An linear chain CRF is defined to maximize the following likelihood function:
+
+    $$ L(W, U, b; y_1, ..., y_n) := \frac{1}{Z} \sum_{y_1, ..., y_n} \exp(-a_1' y_1 - a_n' y_n
+        - \sum_{k=1^n}(((x_k' W + b) y_k) + y_1' U y_2)), $$
+
+    where:
+        $Z$: normalization constant
+        $x_k, y_k$:  inputs and outputs
+
+    This implementation has two modes for optimization:
+    1. (`join mode`) optimized by maximizing join likelihood, which is optimal in theory of statistics.
+       Note that in this case, CRF mast be the output/last layer.
+    2. (`marginal mode`) return marginal probabilities on each time step and optimized via composition
+       likelihood (product of marginal likelihood), i.e., using `categorical_crossentropy` loss.
+       Note that in this case, CRF can be either the last layer or an intermediate layer.
+
+    For prediction (test phrase), one can choose either Viterbi best path (class indices) or marginal
+    probabilities if probabilities are needed. However, if one chooses *join mode* for training,
+    Viterbi output is typically better than marginal output, but the marginal output will still perform
+    reasonably close, while if *marginal mode* is used for training, marginal output usually performs
+    much better. The default behavior is set according to this observation.
+
+    In addition, this implementation supports masking and accepts either onehot or sparse target.
+
+
+    # Examples
+
+    ```python
+        model = Sequential()
+        model.add(Embedding(3001, 300, mask_zero=True)(X)
+
+        # use learn_mode = 'join', test_mode = 'viterbi'
+        crf = CRF(10)
+        model.add(crf(Embed))
```

Comment:
hi @linxihui, 'Embed' seems to be undefined.
I think this line should be ```model.add(crf)``` (followed by [test_crf.py#L34](https://github.com/linxihui/keras/blob/7e8495c5fe4597374ff7b84afe5bbffc9692de75/tests/keras/layers/test_crf.py#L34))